### ISSUE TITLE
New grammar cannot parse

### ISSUE THREAD

**Issue by tathanhdinh** [kevinmehall/rust-peg#207]
I'm trying to convert this [grammar](https://github.com/CraneStation/simplejit-demo/blob/master/src/grammar.rustpeg):

```peg
use super::Expr;

pub function -> (String, Vec<String>, String, Vec<Expr>)
    = [ \t\n]* "fn" _ name: identifier _
      "(" params:((_ i:identifier _ {i}) ** ",") ")" _
      "->" _
      "(" returns:(_ i:identifier _ {i}) ")" _
      "{" _ "\n"
      stmts:statements
      _ "}" _ "\n" _
      { (name, params, returns, stmts) }

statements -> Vec<Expr>
    = s:(statement*) { s }

statement -> Expr
    = _ e:expression _ "\n" { e }

expression -> Expr
    = if_else
    / while_loop
    / i:identifier _ "=" _ e:expression { Expr::Assign(i, Box::new(e)) }
    / compare

if_else -> Expr
    = "if" _ e:expression _ "{" _ "\n"
      then_body:statements _ "}" _ "else" _ "{" _ "\n"
      else_body:statements _ "}"
      { Expr::IfElse(Box::new(e), then_body, else_body) }

while_loop -> Expr
    = "while" _ e:expression _ "{" _ "\n"
      loop_body:statements _ "}"
      { Expr::WhileLoop(Box::new(e), loop_body) }

compare -> Expr
    = a:sum _ "==" _ b:compare { Expr::Eq(Box::new(a), Box::new(b)) }
    / a:sum _ "!=" _ b:compare { Expr::Ne(Box::new(a), Box::new(b)) }
    / a:sum _ "<"  _ b:compare { Expr::Lt(Box::new(a), Box::new(b)) }
    / a:sum _ "<=" _ b:compare { Expr::Le(Box::new(a), Box::new(b)) }
    / a:sum _ ">"  _ b:compare { Expr::Gt(Box::new(a), Box::new(b)) }
    / a:sum _ ">=" _ b:compare { Expr::Ge(Box::new(a), Box::new(b)) }
    / sum

sum -> Expr
    = a:product _ "+" _ b:sum { Expr::Add(Box::new(a), Box::new(b)) }
    / a:product _ "-" _ b:sum { Expr::Sub(Box::new(a), Box::new(b)) }
    / product

product -> Expr
    = a:call_or_identifier_or_literal _ "*" _ b:product { Expr::Mul(Box::new(a), Box::new(b)) }
    / a:call_or_identifier_or_literal _ "/" _ b:product { Expr::Div(Box::new(a), Box::new(b)) }
    / call_or_identifier_or_literal

call_or_identifier_or_literal -> Expr
    = i:identifier _ "(" args:((_ e:expression _ {e}) ** ",") ")" { Expr::Call(i, args) }
    / i:identifier { Expr::Identifier(i) }
    / literal

identifier -> String
    = n:$([a-zA-Z_][a-zA-Z0-9_]*) { n.to_owned() }

literal -> Expr
    = n:$([0-9]+) { Expr::Literal(n.to_owned()) }
    / "&" i:identifier { Expr::GlobalDataAddr(i) }

_ = [ \t]*
```

which is well parsed with `peg v0.5.4`, using the [new syntax](https://github.com/kevinmehall/rust-peg/releases):

```peg
peg::parser!{pub grammar parser() for str {
    use super::Expr;

    pub rule function() -> (String, Vec<String>, String, Vec<Expr>)
        = [' ' | '\t' | '\n']* "fn" _ name:identifier _
        "(" params:((_ i:identifier _ {i}) ** ",") ")" _
        "->" _
        "(" returns:(_ i:identifier _ {i}) ")" _
        "{" _ "\n"
        stmts:statements
        _ "}" _ "\n" _
        { (name, params, returns, stmts) }

    rule statements() -> Vec<Expr>
        = s:(statement*) { s }

    rule statement() -> Expr
        = _ e:expression _ "\n" { e }

    rule expression() -> Expr
        = if_else
        / while_loop
        / i:identifier _ "=" _ e:expression { Expr::Assign(i, Box::new(e)) }
        / compare

    rule if_else() -> Expr
        = "if" _ e:expression _ "{" _ "\n"
        then_body:statements _ "}" _ "else" _ "{" _ "\n"
        else_body:statements _ "}"
        { Expr::IfElse(Box::new(e), then_body, else_body) }

    rule while_loop() -> Expr
        = "while" _ e:expression _ "{" _ "\n"
        loop_body:statements _ "}"
        { Expr::WhileLoop(Box::new(e), loop_body) }

    rule compare() -> Expr
        = a:sum _ "==" _ b:compare { Expr::Eq(Box::new(a), Box::new(b)) }
        / a:sum _ "!=" _ b:compare { Expr::Ne(Box::new(a), Box::new(b)) }
        / a:sum _ "<"  _ b:compare { Expr::Lt(Box::new(a), Box::new(b)) }
        / a:sum _ "<=" _ b:compare { Expr::Le(Box::new(a), Box::new(b)) }
        / a:sum _ ">"  _ b:compare { Expr::Gt(Box::new(a), Box::new(b)) }
        / a:sum _ ">=" _ b:compare { Expr::Ge(Box::new(a), Box::new(b)) }
        / sum

    rule sum() -> Expr
        = a:product _ "+" _ b:sum { Expr::Add(Box::new(a), Box::new(b)) }
        / a:product _ "-" _ b:sum { Expr::Sub(Box::new(a), Box::new(b)) }
        / product

    rule product() -> Expr
        = a:call_or_identifier_or_literal _ "*" _ b:product { Expr::Mul(Box::new(a), Box::new(b)) }
        / a:call_or_identifier_or_literal _ "/" _ b:product { Expr::Div(Box::new(a), Box::new(b)) }
        / call_or_identifier_or_literal

    rule call_or_identifier_or_literal() -> Expr
        = i:identifier _ "(" args:((_ e:expression _ {e}) ** ",") ")" { Expr::Call(i, args) }
        / i:identifier { Expr::Identifier(i) }
        / literal

    rule identifier() -> String
        = n:$(['a'..='z'|'A'..='Z'|'_']['a'..='z'|'A'..='Z'|'0'..='9'|'_']*) { n.to_owned() }

    rule literal() -> Expr
        = n:$(['0'..='9']+) { Expr::Literal(n.to_owned()) }
        / "&" i:identifier { Expr::GlobalDataAddr(i) }

    rule _() = ['\t']*
}}
```
but it didn't parse.

The compiler complains that it expects a `"("` at the `_` rule after identifier

```peg
peg::parser!{pub grammar parser() for str {
    use super::Expr;

    pub rule function() -> (String, Vec<String>, String, Vec<Expr>)
        = [' ' | '\t' | '\n']* "fn" _ name:identifier _   <= expected "(" here
```

But I still don't understand why.

Many thanks for any help.

**Comment by tathanhdinh** [comment]
I found it, all rules must come with `()`.

**Comment by kevinmehall** [comment]
In fact, I wrote the upgrade instructions in the release notes while upgrading that exact grammar, since I'd already upgraded my own projects weeks ago and wanted a final check before releasing 0.6 (and because I've been wanting to play with Cranelift). I submitted [a PR with my changes](https://github.com/CraneStation/simplejit-demo/pull/19). Apologies if you were in the middle of doing the same.
